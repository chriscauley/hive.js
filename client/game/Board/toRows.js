/* Renders the board into a state that can be displayed */
import { range } from 'lodash'
import pieces from '../pieces'
import Board from './index'
import webs from './webs'

const _class = (player, type) => `piece hex-player_${player} type type-${type} hex `
const empty = (board, index) => {
  const cls = 'piece hex'
  if (!board.empty[index]) {
    // hexes far away from the hive are "invisible"
    return cls
  }
  const colors = ['r', 'g', 'b', 'r']
  const r_i = (board.geo.index2xy(index)[1] % 3) + (index % 2)
  return cls + ' hex-empty_' + colors[r_i]
}

const pieceToClass = (board, piece_id) => {
  const type = board.piece_types[piece_id]
  const player = board.piece_owners[piece_id]
  return _class(player, type)
}

const sliceBoard = (board) => {
  let max_x = -Infinity,
    max_y = -Infinity,
    min_x = Infinity,
    min_y = Infinity
  const indexes = Object.keys(board.stacks)
  if (indexes.length === 0) {
    indexes.push(board.geo.center)
  }
  indexes
    .map((index) => parseInt(index))
    .forEach((index) => {
      const xy = board.geo.index2xy(index)
      max_x = Math.max(max_x, xy[0] + 1)
      max_y = Math.max(max_y, xy[1] + 1)
      min_x = Math.min(min_x, xy[0] - 1)
      min_y = Math.min(min_y, xy[1] - 1)
    })
  if (min_x % 2 !== 0) {
    min_x -= 1
  }
  return range(min_y, max_y + 1).map((y) =>
    range(min_x, max_x + 1).map((x) => ({
      index: board.geo.xy2index([x, y]),
      xy: [x, y],
    })),
  )
}

export default (board, { columns = 1 } = {}) => {
  const marked = getMarked(board)
  const { selected = {} } = board
  const show_webs = webs.getVisible(board, selected.index)
  if (selected.index !== undefined) {
    marked[selected.index] = ' purple'
  } else if (board.last) {
    if (board.last.from) {
      marked[board.last.from] = ' blue-dashed'
    }
    if (board.last.to) {
      marked[board.last.to] = ' blue'
    }
    if (board.last.special) {
      marked[board.last.special] = ' yellow'
    }
  }
  const rows = sliceBoard(board)
  rows.forEach((row) =>
    row.forEach((cell) => {
      cell.stack = []
      cell.selected = cell.index === selected.index
      if (board.stacks[cell.index]) {
        board.stacks[cell.index].forEach((piece_id) => {
          cell.stack.push(pieceToClass(board, piece_id))
          // last piece_id gets used here
          cell.piece_id = piece_id
          cell.player_id = board.piece_owners[piece_id]
          cell.piece_type = board.piece_types[piece_id]
        })
      } else {
        cell.stack.push(empty(board, cell.index))
      }
      show_webs.forEach((web) => {
        if (!board.layers[web][cell.index]) {
          return // no webs
        }
        if (web === 'crawl') {
          if (board.stacks[cell.index] && cell.index !== selected.index) {
            // only show crawl web on the selected piece or empty squares
            return
          }
          const enemy_webs = board.layers[web][cell.index].filter(
            (i) => selected.player_id !== board.layers.player[i],
          )
          if (enemy_webs.length === 0) {
            return
          }

          const selected_webs = board.layers.crawl[selected.index] || []
          if (enemy_webs.find((i) => selected_webs.includes(i)) !== undefined) {
            web = 'crawl-gray'
          }
        }
        cell.title = webs.title[web](selected.piece_type)
        cell.web = web
      })
      if (marked[cell.index]) {
        const _i = cell.stack.length - 1
        cell.stack[_i] = cell.stack[_i] + marked[cell.index]
      }
    }),
  )

  const players = {
    1: [],
    2: [],
  }

  pieces.getAvailable(board).forEach(({ player_id, type, count }) => {
    player_id = parseInt(player_id)
    const className = _class(player_id, type)
    const cell = {
      stack: range(count).map(() => className),
      player_id,
      piece_id: 'new',
      piece_type: type, // TODO remove drag and drop and then this can be type
      type: 'cell',
    }
    if (
      selected.piece_id === 'new' &&
      selected.piece_type === type &&
      player_id === selected.player_id
    ) {
      const _i = cell.stack.length - 1
      cell.stack[_i] = cell.stack[_i] + ' purple'
    }
    players[player_id].push(cell)
  })

  return {
    rows,
    player_1: columnize(players[1], columns),
    player_2: columnize(players[2], columns),
  }
}

const columnize = (cells, columns) => {
  const out = []
  let row = []
  out.push(row)
  while (cells.length) {
    row.push(cells.shift())
    if (row.length === columns) {
      row = []
      out.push(row)
    }
  }
  return out
}

const getMarked = (board) => {
  const out = {}
  Object.keys(board.cantmove).forEach((index) => (out[index] = 'gray'))

  if (!board.selected) {
    return out
  }

  const { piece_id, player_id } = board.selected

  const color = board.current_player === player_id ? ' green' : ' red'
  let indexes = []
  if (piece_id === 'new') {
    indexes = Board.moves.getPlacement(board, player_id)
  } else {
    const specials = Board.getSpecials(board, piece_id, board.special_args)
    specials.forEach((i) => (out[i] = ' yellow'))
    if (board.special_args.length === 0) {
      indexes = Board.getMoves(board, piece_id)
    }
  }
  indexes.forEach((i) => (out[i] = color))
  if (board.piece_types[piece_id] === 'dragonfly') {
    const index = board.reverse[piece_id]
    indexes.forEach((i2) => (out[i2] += Board.moves.dragonflyExtra(board, index, i2)))
  }
  return out
}
