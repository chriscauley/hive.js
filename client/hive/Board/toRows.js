import { getGeo } from './Geo'

const pieceToClass = (board, piece_id) => {
  const type = board.piece_types[piece_id]
  const player = board.piece_owners[piece_id]
  return `piece hex-player_${player} type type-${type} hex `
}

export default (board) => {
  const used = {}
  const rows = []
  let row
  getGeo(board).indexes.forEach((index) => {
    if (index % board.W === 0) {
      row = []
      rows.push(row)
    }
    const cell = {
      index,
      stack: [],
      type: 'cell',
    }
    row.push(cell)
    board.stacks[index] &&
      board.stacks[index].forEach((piece_id) => {
        used[piece_id] = true
        cell.stack.push(pieceToClass(board, piece_id))
        cell.piece_id = piece_id // last piece_id gets used here
      })
  })

  const players = {
    1: [],
    2: [],
  }

  board.piece_owners.forEach((owner, piece_id) => {
    !used[piece_id] &&
      players[owner].push([
        {
          stack: [pieceToClass(board, piece_id)],
          piece_id,
          type: 'cell',
        },
      ])
  })
  return {
    rows,
    player_1: players[1],
    player_2: players[2],
  }
}
