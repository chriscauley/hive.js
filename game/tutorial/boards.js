/* eslint-disable */
// not linting becauset his file should mostly be treated like binary

export default {
  queen: {"W":8,"H":6,"stacks":{"18":[2],"20":[4],"25":[7],"27":[5],"29":[1],"35":[3],"36":[6],"37":[0]},"piece_types":["queen","beetle","queen","ant","spider","ant","ant","grasshopper"],"piece_owners":[1,2,2,2,1,1,1,2],"turn":20,"rules":{}, "current_player": 1},
  beetle: {"W":6,"H":5,"stacks":{"8":[1],"13":[0,2],"14":[6],"15":[5],"20":[3],"21":[4]},"piece_types":["grasshopper","queen","beetle","queen","grasshopper","beetle","grasshopper"],"piece_owners":[2,2,1,1,1,2,2],"turn":9,"rules":{}, "current_player": 2},
  grasshopper: {"W":8,"H":4,"stacks":{"10":[2],"11":[0],"12":[1],"18":[4],"19":[3],"20":[5],"21":[6]},"piece_types":["queen","beetle","grasshopper","ant","queen","spider","grasshopper"],"piece_owners":[2,2,2,1,1,1,1],"turn":9,"rules":{}, "current_player": 2},
  spider: {"W":8,"H":6,"stacks":{"19":[0],"20":[1],"26":[5],"27":[3],"28":[4],"29":[2]},"piece_types":["spider","beetle","queen","queen","beetle","grasshopper"],"piece_owners":[1,1,1,2,2,2],"turn":13,"rules":{}, "current_player": 2},
  ant: {"W":8,"H":6,"stacks":{"12":[7],"18":[5],"20":[1],"25":[6],"27":[2],"28":[4],"29":[0],"35":[8],"37":[3]},"piece_types":["ant","spider","grasshopper","grasshopper","grasshopper","spider","queen","queen","ant"],"piece_owners":[2,1,1,2,2,1,1,2,1],"turn":42,"rules":{}, "current_player": 1},
  ladybug: {"W":6,"H":5,"stacks":{"7":[0],"9":[5],"10":[8],"13":[1],"14":[2],"15":[4],"16":[6],"19":[7],"20":[3]},"piece_types":["ladybug","grasshopper","beetle","queen","mosquito","queen","spider","ant","ladybug"],"piece_owners":[2,1,1,1,2,2,1,2,1],"turn":10,"rules":{}, "current_player": 1},
  mosquito: {"W":6,"H":6,"stacks":{"14":[0],"15":[3],"16":[5],"20":[1,6],"21":[2],"22":[4]},"piece_types":["scorpion","queen","grasshopper","mosquito","beetle","queen","mosquito"],"piece_owners":[2,1,1,2,1,2,1],"turn":16,"rules":{}, "current_player": 1},
  pill_bug: {"W":6,"H":6,"stacks":{"8":[4],"14":[3],"15":[5],"19":[0],"20":[2],"25":[1]},"piece_types":["pill_bug","queen","spider","grasshopper","queen","ant"],"piece_owners":[1,1,1,2,2,2],"turn":8,"rules":{},"current_player": 1},
  orchid_mantis: {"W":8,"H":6,"stacks":{"11":[5,6],"12":[3],"20":[2],"21":[4],"27":[0],"28":[1],"35":[7]},"piece_types":["orchid_mantis","queen","beetle","queen","ant","ant","orchid_mantis","grasshopper"],"piece_owners":[1,1,2,2,2,1,2,1],"turn":9,"rules":{}, "current_player": 2},
  fly: {"W":50,"H":50,"stacks":{"1176":[6],"1225":[5],"1227":[8],"1273":[4],"1274":[0,2],"1275":[1],"1276":[3],"1277":[7]},"piece_types":["queen","queen","fly","orchid_mantis","wasp","fly","wasp","scorpion","scorpion"],"piece_owners":[1,2,2,2,1,1,1,2,1],"turn":14,"rules":{},"current_player":1},
  scorpion: {"W":6,"H":5,"stacks":{"8":[2],"13":[1],"15":[3],"16":[5],"19":[0],"21":[4],"22":[6]},"piece_types":["queen","spider","beetle","queen","grasshopper","beetle","scorpion"],"piece_owners":[1,1,2,2,2,2,1],"turn":11,"rules":{}, "current_player": 2},
  wasp: {"W":6,"H":6,"stacks":{"8":[1],"13":[0],"14":[2],"15":[4],"19":[3],"20":[5],"25":[6]},"piece_types":["ant","queen","wasp","queen","wasp","beetle","grasshopper"],"piece_owners":[2,2,2,1,1,1,2],"turn":8,"rules":{}, "current_player": 1},
  dragonfly: {"W":6,"H":5,"stacks":{"7":[7],"8":[3],"14":[1,0],"15":[4],"19":[6],"20":[2],"21":[5]},"piece_types":["dragonfly","spider","queen","scorpion","queen","beetle","dragonfly","spider"],"piece_owners":[1,1,1,2,2,2,2,1],"turn":14,"rules":{}, "current_player": 1},
  damselfly: {"W":6,"H":5,"stacks":{"7":[7],"8":[3],"14":[1,0],"15":[4],"19":[6],"20":[2],"21":[5]},"piece_types":["damselfly","spider","queen","scorpion","queen","beetle","damselfly","spider"],"piece_owners":[1,1,1,2,2,2,2,1],"turn":14,"rules":{}, "current_player": 1},
  dragonfly_nymph: {"actions":[],"W":50,"H":50,"stacks":{"1225":[4],"1273":[2],"1274":[3,0],"1275":[1],"1323":[5]},"piece_types":["dragonfly_nymph","dragonfly_nymph","queen","queen","lanternfly_nymph","centipede"],"piece_owners":[1,2,1,2,2,1],"hash":"cb6ca89c583ff67546d37a5b4453bf1b714fe20e","turn":7,"rules":{},"last":{"to":1323},"current_player":2},
  centipede: {"W":6,"H":6,"stacks":{"10":[6],"14":[4],"15":[3,5],"20":[0],"21":[2],"27":[1]},"piece_types":["queen","grasshopper","centipede","ant","ladybug","beetle","queen"],"piece_owners":[1,1,2,2,2,1,2],"turn":10,"rules":{}, "current_player": 1},
  cockroach: {"W":8,"H":6,"stacks":{"12":[3],"14":[6],"18":[1],"19":[2],"20":[5],"21":[4,7],"27":[0]},"piece_types":["queen","cockroach","beetle","queen","fly","ant","cockroach","beetle"],"piece_owners":[1,1,2,2,2,1,2,1],"hash":0.5479472595173172,"turn":20,"rules":{}, "current_player": 1},
  cicada: {"actions":[],"W":50,"H":50,"stacks":{"1226":[3],"1273":[2],"1274":[0],"1275":[1],"1323":[4]},"piece_types":["queen","queen","cicada","cicada","ant"],"piece_owners":[1,2,1,2,1],"hash":"21f05bca4e8cc4257c4de80ac6dfdd6aaa971ec0","turn":5,"rules":{"pieces":{"beetle":2,"ant":3,"centipede":1,"cockroach":1,"cicada":2}},"current_player":2},
  lanternfly: {"W":50,"H":50,"stacks":{"1222":[6],"1225":[5],"1226":[7],"1273":[2],"1274":[0],"1275":[1],"1276":[3],"1323":[4]},"piece_types":["queen","beetle","beetle","queen","lanternfly","ant","cicada","lanternfly"],"piece_owners":[1,2,1,2,1,2,1,2],"hash":"c3ebd4ae845992bc447b61ee23449c23cfd8ea41","turn":8,"rules":{"pieces":{"beetle":2,"ant":3,"lanternfly":1,"centipede":1,"cockroach":1,"cicada":2}},"current_player":1},
  lanternfly_nymph: {"actions":[],"W":50,"H":50,"stacks":{"1223":[8],"1226":[5],"1273":[2],"1274":[0],"1275":[1],"1276":[3],"1323":[4],"1324":[6],"1326":[7]},"piece_types":["pill_bug","mosquito","queen","pill_bug","lanternfly_nymph","lanternfly_nymph","centipede","queen","orchid_mantis"],"piece_owners":[1,2,1,2,1,2,1,2,1],"hash":"3e89a125b257082f0a7cd6687cec8c0303eab5d3","turn":16,"rules":{},current_player: 1},
  trapdoor_spider: {"W":50,"H":50,"stacks":{"1226":[6],"1274":[0],"1275":[1],"1276":[3],"1277":[5],"1323":[4],"1325":[2],"1327":[7]},"piece_types":["trapdoor_spider","trapdoor_spider","queen","queen","cicada","ant","ant","cicada"],"piece_owners":[1,2,1,2,1,2,1,2],"rules": {},"current_player":2},
  orbweaver: {"actions":[],"W":50,"H":50,"stacks":{"1174":[10,8],"1176":[5],"1222":[12],"1225":[3],"1272":[6],"1273":[2],"1274":[0],"1275":[1],"1276":[7],"1325":[11],"1322":[9],"1323":[4]},"piece_types":["orbweaver","grasshopper","grasshopper","orbweaver","queen","queen","wasp","fly","fly","trapdoor_spider","trapdoor_spider","cockroach","cockroach"],"piece_owners":[1,2,1,2,1,2,1,2,1,2,1,2,1],"turn":18,"rules":{},"last":{"from":1226,"to":1277},"current_player":1,"last_move_at":1604858870793},
  praying_mantis: {"W":50,"H":50,"stacks":{"1225":[5],"1273":[6],"1274":[0,4],"1275":[1],"1276":[7,3],"1323":[2]},"piece_types":["queen","queen","praying_mantis","praying_mantis","beetle","fly","ant","beetle"],"piece_owners":[1,2,1,2,1,2,1,2],"turn":14,"last":{"from":1325,"special":1276},"current_player":1, rules: {}},
  emerald_wasp: {"W":50,"H":50,"stacks":{"1225":[5],"1273":[2],"1274":[0],"1275":[1,4],"1276":[3]},"piece_types":["beetle","beetle","queen","queen","emerald_wasp","emerald_wasp"],"piece_owners":[1,2,1,2,1,2],"turn":7,"rules":{},"current_player": 2},
  kung_fu_mantis: {"W":50,"H":50,"stacks":{"1224":[7],"1225":[4,3],"1274":[0],"1275":[1],"1276":[5],"1323":[6],"1324":[2],"1325":[8]},"piece_types":["queen","queen","ant","kung_fu_mantis","grasshopper","ant","kung_fu_mantis","ladybug","ladybug"],"piece_owners":[1,2,1,2,1,2,1,1,2],"turn":13,"rules":{},"current_player":2},
  earthworm: {"actions":[],"W":50,"H":50,"stacks":{"1223":[8],"1225":[3],"1226":[7],"1273":[0,6],"1274":[2],"1275":[1],"1276":[5],"1323":[4]},"piece_types":["centipede","pill_bug","queen","earthworm","earthworm","dragonfly","beetle","queen","lanternfly_nymph"],"piece_owners":[1,2,1,2,1,2,1,2,1],"hash":"1d4cc628d73f7cc48ab20a7b3a4800f37e13b3f3","turn":16,"rules":{},"current_player":1},
}