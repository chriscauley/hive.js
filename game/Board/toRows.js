// TODO will be toCells afterwards
/* Renders the board into a state that can be displayed */
import { range } from 'lodash'
import pieces from '../pieces'
import Board from './index'
import webs from './webs'

const COLORS = ['r', 'g', 'b', 'r']

const empty = (board, index) => {
  const r_i = (board.geo.index2xy(index)[1] % 3) + (index % 2)
  return `piece hex hex-empty_${COLORS[r_i]}`
}

const sliceBoard = (board) => {
  let max_x = -Infinity,
    max_y = -Infinity,
    min_x = Infinity,
    min_y = Infinity
  const indexes = Object.keys(board.stacks)
  if (indexes.length === 0) {
    indexes.push(board.geo.center)
  }
  indexes
    .map((index) => parseInt(index))
    .forEach((index) => {
      const xy = board.geo.index2xy(index)
      max_x = Math.max(max_x, xy[0] + 1)
      max_y = Math.max(max_y, xy[1] + 1)
      min_x = Math.min(min_x, xy[0] - 1)
      min_y = Math.min(min_y, xy[1] - 1)
    })
  if (min_x % 2 !== 0) {
    min_x -= 1
  }
  return range(min_y, max_y + 1).map((y) =>
    range(min_x, max_x + 1).map((x) => ({
      index: board.geo.xy2index([x, y]),
      xy: [x, y],
    })),
  )
}

export const makeStack = (board, index) => {
  const stack = board.stacks[index]
  const xy = board.geo.index2xy(index)
  if (board.empty[index]) {
    return [
      {
        xy,
        index,
        class: [empty(board, index)],
        z: 0,
        id: index,
      },
    ]
  }
  if (!stack) {
    return
  }
  return stack.map((piece_id, stack_index) => {
    const z_hover = stack.length - stack_index
    let z = z_hover
    if (stack.length > 4) {
      z = z_hover - stack.length + 4
    }
    const player_id = board.piece_owners[piece_id]
    const piece_type = board.piece_types[piece_id]
    return {
      class: [
        `hex hex-player_${player_id} type type-${piece_type} piece`,
        `-z-${z} -z-hover-${z_hover}`,
      ],
      player_id,
      piece_type,
      index,
      xy,
      id: piece_id,
    }
  })
}

export default (board) => {
  const marked = getMarked(board)
  const { selected = {} } = board
  if (selected.index !== undefined) {
    marked[selected.index] = ' purple' // TOOD note 1
  } else if (selected.piece_id !== 'new' && board.last) {
    if (board.last.from) {
      marked[board.last.from] = ' blue-dashed'
    }
    if (board.last.to) {
      marked[board.last.to] = ' blue'
    }
    if (board.last.special) {
      marked[board.last.special] = ' yellow'
    }
    board.last.stacks?.forEach((index) => {
      marked[index] += ' purple-inner'
    })
  }

  const show_webs = webs.getVisible(board, selected.index)
  const rows = sliceBoard(board)
  rows.forEach((row) =>
    row.forEach((cell) => {
      cell.stack = makeStack(board, cell.index)
      if (!cell.stack) {
        return
      }
      const top = (cell.top = cell.stack[cell.stack.length - 1])
      top.top = true
      if (cell.index === selected.index) {
        top.selected = true
        top.class.push('purple') // TODO does marked cover this? (note 1)
      }
      const stack = board.stacks[cell.index]
      if (stack) {
        const piece_id = stack[stack.length - 1]
        cell.piece_id = piece_id
        cell.player_id = board.piece_owners[piece_id]
        cell.piece_type = board.piece_types[piece_id]
      }
      show_webs.forEach((web) => {
        if (!board.layers[web][cell.index]) {
          return // no webs
        }
        if (web === 'stack' && selected.player_id === board.layers.player[cell.index]) {
          return
        }
        if (web === 'crawl') {
          if (board.stacks[cell.index] && cell.index !== selected.index) {
            // only show crawl web on the selected piece or empty squares
            return
          }
          const enemy_webs = board.layers[web][cell.index].filter(
            (i) => selected.player_id !== board.layers.player[i],
          )
          if (enemy_webs.length === 0) {
            return
          }

          const selected_webs = board.layers.crawl[selected.index] || []
          if (enemy_webs.find((i) => selected_webs.includes(i)) !== undefined) {
            web = 'crawl-gray'
          }
        }
        cell.title = webs.title[web](selected.piece_type)
        cell.web = web
        top.class.push(web)
      })
      if (marked[cell.index]) {
        top.class.push(marked[cell.index])
      }
    }),
  )

  const players = {
    1: [],
    2: [],
  }

  pieces.getAvailable(board).forEach(({ player_id, type, count }) => {
    let index = players[player_id].length
    if (index !== 0) {
      index++ // looks better in player piece bins
    }
    const xy = [index % 2, Math.floor(index / 2)]
    player_id = parseInt(player_id)
    const cell = {
      stack: range(count).map((z) => ({
        class: [`hex hex-player_${player_id} type type-${type} piece -z-${z}`],
        player_id,
        piece_type: type,
        index,
        xy,
        id: `new_${player_id}_${type}_${z}`,
      })),
      index,
      player_id,
      piece_id: 'new',
      piece_type: type, // TODO remove drag and drop and then this can be type
      type: 'cell',
      xy,
    }
    if (
      selected.piece_id === 'new' &&
      selected.piece_type === type &&
      player_id === selected.player_id
    ) {
      const _i = cell.stack.length - 1
      cell.stack[cell.stack.length].class.push('purple')
    }
    players[player_id].push(cell)
  })

  return {
    rows,
    player_1: [players[1]],
    player_2: [players[2]],
  }
}

const getMarked = (board) => {
  const { selected } = board
  const marked = {}
  Object.keys(board.cantmove).forEach((index) => (marked[index] = ' gray'))

  if (!selected) {
    return marked
  }

  const { piece_id, player_id } = selected

  const color = board.current_player === player_id ? ' green' : ' red'
  let indexes = []
  if (piece_id === 'new') {
    indexes = Board.moves.getPlacement(board, player_id)
  } else {
    const specials = Board.getSpecials(board, piece_id, board.special_args)
    specials.forEach((i) => (marked[i] = ' yellow'))
    if (board.special_args.length === 0) {
      indexes = Board.getMoves(board, piece_id)
    }
    if (selected.chalk) {
      Object.entries(selected.chalk).forEach(
        ([index, f]) => (marked[index] = f(marked[index] || '')),
      )
    }
  }
  indexes.forEach((i) => (marked[i] = color))
  return marked
}
