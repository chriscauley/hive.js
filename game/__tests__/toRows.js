import toRows, { makeStack } from 'hive.js/Board/toRows'
import boards from 'hive.js/tutorial/boards'
import B from 'hive.js/Board'
import { cloneDeep, last } from 'lodash'

const stack_board = {
  "W":50,
  "H":50,
  "stacks":{
    "1226":[4,7,8],
    "1275":[0,1,2,5,3,6],
    "1276":[9]
  },
  "piece_types": [0,1,2,3,4,5,6,7,8,9],
  "piece_owners":[2,1,1,1,2,1,1,2,2,2],
  "rules":{},
  "current_player":1,
}


test('makeStack', () => {
  const board = cloneDeep(stack_board)
  B.update(board)
  expect(JSON.stringify(toRows(board))).toMatchSnapshot()

  const empty_board = cloneDeep(stack_board)
  empty_board.stacks = {}
  empty_board.piece_owners = empty_board.piece_types = []
  B.update(empty_board)
  toRows(empty_board)
})

const moves_board = {"W":50,"H":50,"stacks":{"1224":[5],"1226":[8],"1227":[10],"1274":[0],"1275":[1],"1276":[6],"1323":[3],"1324":[2],"1325":[4],"1326":[7],"1327":[9]},"piece_types":["trapdoor_spider","trapdoor_spider","scorpion","praying_mantis","praying_mantis","queen","queen","cicada","orbweaver","ant","beetle"],"piece_owners":[2,1,2,2,1,2,1,2,1,2,2],"hash":"cd140fbc0ec2c520a1e082271d8580083583bbab","turn":15,"rules":{"pieces":{"ant":1,"queen":1,"beetle":2,"cicada":1,"spider":0,"scorpion":1,"orbweaver":1,"grasshopper":0,"praying_mantis":1,"dragonfly_nymph":1,"trapdoor_spider":1},"variants":{},"no_rules":true},"players":{"1":407,"2":404},"last":{"to":1227},"current_player":1,"last_move_at":1624304370413}

test.only('toRows', () => {
  const board = cloneDeep(moves_board)
  // Object.keys(board.stacks)
  ;[1326]
    .forEach((index) => {
    index = Number(index)
    const piece_id = last(board.stacks[index])
    const piece_type = board.piece_types[piece_id]
    const player_id = board.piece_owners[piece_id]
    B.update(board)
    B.select(board, { index, piece_id, piece_type, player_id })
    toRows(board)
  })
})