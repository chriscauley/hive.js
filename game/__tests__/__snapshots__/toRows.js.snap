// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeStack 1`] = `"{\\"rows\\":[[{\\"index\\":1174,\\"xy\\":[24,23],\\"stack\\":[\\"piece hex\\"],\\"selected\\":false},{\\"index\\":1175,\\"xy\\":[25,23],\\"stack\\":[\\"piece hex\\"],\\"selected\\":false},{\\"index\\":1176,\\"xy\\":[26,23],\\"stack\\":[\\"piece hex hex-empty_b\\"],\\"selected\\":false},{\\"index\\":1177,\\"xy\\":[27,23],\\"stack\\":[\\"piece hex\\"],\\"selected\\":false}],[{\\"index\\":1224,\\"xy\\":[24,24],\\"stack\\":[\\"piece hex hex-empty_r\\"],\\"selected\\":false},{\\"index\\":1225,\\"xy\\":[25,24],\\"stack\\":[\\"piece hex hex-empty_g\\"],\\"selected\\":false},{\\"index\\":1226,\\"xy\\":[26,24],\\"stack\\":[\\"piece hex-player_2 type type-4 hex -stacked-0\\",\\"piece hex-player_2 type type-7 hex -stacked-1\\",\\"piece hex-player_2 type type-8 hex -stacked-2\\"],\\"selected\\":false,\\"piece_id\\":8,\\"player_id\\":2,\\"piece_type\\":8},{\\"index\\":1227,\\"xy\\":[27,24],\\"stack\\":[\\"piece hex hex-empty_g\\"],\\"selected\\":false}],[{\\"index\\":1274,\\"xy\\":[24,25],\\"stack\\":[\\"piece hex hex-empty_g\\"],\\"selected\\":false},{\\"index\\":1275,\\"xy\\":[25,25],\\"stack\\":[\\"piece hex-player_2 type type-0 hex -stacked--2\\",\\"piece hex-player_1 type type-1 hex -stacked--1\\",\\"piece hex-player_1 type type-2 hex -stacked-0\\",\\"piece hex-player_1 type type-5 hex -stacked-1\\",\\"piece hex-player_1 type type-3 hex -stacked-2\\",\\"piece hex-player_1 type type-6 hex -stacked-3\\"],\\"selected\\":false,\\"piece_id\\":6,\\"player_id\\":1,\\"piece_type\\":6},{\\"index\\":1276,\\"xy\\":[26,25],\\"stack\\":[\\"piece hex-player_2 type type-9 hex -stacked-0\\"],\\"selected\\":false,\\"piece_id\\":9,\\"player_id\\":2,\\"piece_type\\":9},{\\"index\\":1277,\\"xy\\":[27,25],\\"stack\\":[\\"piece hex hex-empty_b\\"],\\"selected\\":false}],[{\\"index\\":1324,\\"xy\\":[24,26],\\"stack\\":[\\"piece hex\\"],\\"selected\\":false},{\\"index\\":1325,\\"xy\\":[25,26],\\"stack\\":[\\"piece hex hex-empty_r\\"],\\"selected\\":false},{\\"index\\":1326,\\"xy\\":[26,26],\\"stack\\":[\\"piece hex hex-empty_b\\"],\\"selected\\":false},{\\"index\\":1327,\\"xy\\":[27,26],\\"stack\\":[\\"piece hex hex-empty_r\\"],\\"selected\\":false}]],\\"player_1\\":[[{\\"stack\\":[\\"piece hex-player_1 type type-queen hex -stacked-0\\"],\\"player_id\\":1,\\"piece_id\\":\\"new\\",\\"piece_type\\":\\"queen\\",\\"type\\":\\"cell\\"}],[]],\\"player_2\\":[[{\\"stack\\":[\\"piece hex-player_2 type type-queen hex -stacked-0\\"],\\"player_id\\":2,\\"piece_id\\":\\"new\\",\\"piece_type\\":\\"queen\\",\\"type\\":\\"cell\\"}],[]]}"`;
